# Converter grib files to Geotiff

##### Author  : Marcio Teixeira
##### Rev     : PA1 (12-Apr-2022)

### ERA5

In [3]:
# import OS module
import os

# Get the list of all files and directories
path = "/home/mapbiomasar/MarcoFranco/ERA5/Wind_Comp_U_V"
dir_list = os.listdir(path)

print("Files and directories in '", path, "' :")
print(dir_list)


Files and directories in ' /home/mapbiomasar/MarcoFranco/ERA5/Wind_Comp_U_V ' :
['2013.grib', '1994.grib', '2014.grib', '1989.grib', '2021.grib', '2001.grib', '2015.grib', '1999.grib', '2000.grib', '2012.grib', '1986.grib', '1997.grib', '2016.grib', '1988.grib', 'Read_Me.txt', '1990.grib', '1987.grib', '2017.grib', '1991.grib', '2006.grib', '1995.grib', '1979.grib', '1981.grib', '2002.grib', '2008.grib', '1983.grib', '1984.grib', '2018.grib', '2004.grib', '2007.grib', '1996.grib', '1998.grib', '2020.grib', '1985.grib', '1980.grib', '2011.grib', '2019.grib', '2010.grib', '1993.grib', '2003.grib', '1992.grib', '1982.grib', '2009.grib', '2005.grib']


In [6]:
from osgeo import gdal, gdalconst

#Open existing dataset
src_ds = gdal.Open('/home/mapbiomasar/MarcoFranco/ERA5/Wind_Comp_U_V/2013.grib')

#Open output format driver, see gdal_translate --formats for list
format = "GTiff"
driver = gdal.GetDriverByName( format )

#Output to new format
dst_ds = driver.CreateCopy('2013ERA5.tiff', src_ds, 0 )

#Properly close the datasets to flush to disk
dst_ds = None
src_ds = None


ERROR: Ran out of file reading SECT0


ERROR 1: Attempt to create 217x290x70080 TIFF file, but bands must be lesser or equal to 65535.


This is the error of these tiff files

-------
Warning: Inside GRIB2Inventory, Message # 70081
ERROR: Ran out of file reading SECT0
ERROR 1: Attempt to create 217x290x70080 TIFF file, but bands must be lesser or equal to 65535.
-------

### MERGE-CPTEC


It is working! But we will try to make a more recurrent function to process data


In [ ]:
# this is a sandbox

from osgeo import gdal, gdalconst

path = "/home/mapbiomasar/MarcoFranco/MERGE_CPTEC/2000/09"

dir_list = os.listdir(path)
# print(dir_list.sort)
for item in dir_list:
    file_name = item
    #Open existing dataset
    src_ds = gdal.Open(path+'/'+file_name)

    #Open output format driver, see gdal_translate --formats for list
    format = "GTiff"
    driver = gdal.GetDriverByName( format )

    #Output to new format
    dst_ds = driver.CreateCopy(file_name[:-6]+'.tiff', src_ds, 0 )

    #Properly close the datasets to flush to disk
    dst_ds = None
    src_ds = None

    print('File', file_name, 'processed!')
        


### Function that parses the whole directory


In [10]:
from osgeo import gdal, gdalconst


# converts from grib2 to tiff
# TODO: check if converted files are OK

def CptectoTiff(path_to_file,file_):
    src_ds = gdal.Open(path_to_file+file_)

    #Open output format driver, see gdal_translate --formats for list
    format = "GTiff"
    driver = gdal.GetDriverByName( format )

    #Output to new format
    dst_ds = driver.CreateCopy(path_to_file + file_[:-6]+'.tiff', src_ds, 0 )

    #Properly close the datasets to flush to disk
    dst_ds = None
    src_ds = None

# MAIN
# TODO: Make function and more modular as per Python style


path = "/home/mapbiomasar/MarcoFranco/MERGE_CPTEC/"
dir_l1 = os.listdir(path)
# find the files recurrently
for yr in dir_l1:
    if yr.isnumeric():
        print("Processed ", yr, "data")
        year =  path + yr
        dir_l2 = os.listdir(year)
        for mon in dir_l2:
            if mon.isnumeric():
                path_to_file =  year +'/'+ mon + '/'
                dir_l3 = os.listdir(path_to_file)
                for file_ in dir_l3:
                    # path_to_file is the full path
                    # file_ is the *actual* filename
                    CptectoTiff(path_to_file,file_)



Processed  2022 data
Processed  2021 data
Processed  2008 data
Processed  2011 data
Processed  2001 data
Processed  2018 data
Processed  2007 data
Processed  2003 data
Processed  2004 data
Processed  2005 data
Processed  2012 data
Processed  2016 data
Processed  2009 data
Processed  2015 data
Processed  2017 data
Processed  2010 data
Processed  2013 data
Processed  2006 data
Processed  2000 data
Processed  2019 data
Processed  2020 data
Processed  2014 data
Processed  2002 data
